fix(upgrade): fix HMR for hybrid applications

gkalpak opened the pull request:
Previously, trying to apply a change via Hot Module Replacement (HMR) in a hybrid app would result in an error. This was caused by not having the AngularJS app destroyed and thus trying to bootstrap an AngularJS app on the same element twice.

This commit fixes HMR for hybrid apps by ensuring the AngularJS app is destroyed when the Angular `PlatformRef` is [destroyed][1] in the [`module.hot.dispose()` callback][2].

NOTE:
For "ngUpgradeLite" apps (i.e. those using `downgradeModule()`), HMR will only work if the downgraded module has been bootstrapped and there is at least one Angular component present on the page. The is due to a combination of two facts:
- The logic for setting up the listener that destroys the AngularJS app depends on the downgraded module's `NgModuleRef`, which is only available after the module has been bootstrapped.
- The [HMR dispose logic][3] depends on having an Angular element (identified by the auto-geenrated `ng-version` attribute) present in the DOM in order to retrieve the Angular `PlatformRef`.

Fixes #39935.

[1]:
https://github.com/angular/angular-cli/blob/205ea2b638f154291993bfd9e065cd66ff20503/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L75
[2]:
https://github.com/angular/angular-cli/blob/205ea2b638f154291993bfd9e065cd66ff205033/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L31
[3]:
https://github.com/angular/angular-cli/blob/205ea2b638f154291993bfd9e065cd66ff205033/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L116

mary-poppins commented:
You can preview 204e21e at https://pr40045-204e21e.ngbuilds.io/.

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

atscott reviewed (approved):
Reviewed-for: public-api

atscott commented:
[presubmit](https://test.corp.google.com/ui#id=OCL:346563816:BASE:346563845:1607533562624:2523a52a)

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

petebacondarwin commented on packages/upgrade/src/dynamic/src/upgrade_adapter.ts:
```diff
@@ -619,6 +619,12 @@ export class UpgradeAdapter {
+
+                    // Destroy the AngularJS app once the Angular `PlatformRef` is destroyed.
+                    // This does not happen in a typical SPA scenario, but it might be useful for
+                    // other usecases where desposing of an Angular/AngularJS app is necessary (such
```
```suggestion
                    // other use-cases where disposing of an Angular/AngularJS app is necessary (such
```

gkalpak replied:
Ooops 😅 Fixed.

petebacondarwin commented on packages/upgrade/src/dynamic/test/upgrade_spec.ts:
```diff
@@ -86,7 +86,7 @@ withEachNg1Version(() => {
          }));
 
-      it('supports the compilerOptions argument', waitForAsync(() => {
+      it('should support the compilerOptions argument', waitForAsync(() => {
```
😍

petebacondarwin commented on packages/upgrade/static/src/downgrade_module.ts:
```diff
@@ -167,6 +167,12 @@ export function downgradeModule<T>(moduleFactoryOrBootstrapFn: NgModuleFactory<T
 
+                  // Destroy the AngularJS app once the Angular `PlatformRef` is destroyed.
+                  // This does not happen in a typical SPA scenario, but it might be useful for
+                  // other usecases where desposing of an Angular/AngularJS app is necessary (such
```
```suggestion
                  // other use-cases where disposing of an Angular/AngularJS app is necessary (such
```

gkalpak replied:
Ooops 😅 Fixed.

petebacondarwin reviewed (approved):
Can you explain why this is the case?

> For "ngUpgradeLite" apps (i.e. those using `downgradeModule()`), HMR
will only work if there is at least one Angular component present on the
page.

gkalpak committed:
refactor(upgrade): remove unused variables

This commit removes a couple of unused variables.

gkalpak committed:
refactor(upgrade): create a helper for cleaning jqLite/jQuery data

This commit moves the code for cleaning jqLite/jQuery data on an element
to a re-usable helper function. This way it is easier to keep the code
consistent across all places where we need to clean data (now and in the
future).

gkalpak committed:
fixup! refactor(upgrade): create a helper for cleaning jqLite/jQuery data

gkalpak committed:
fix(upgrade): fix HMR for hybrid applications

Previously, trying to apply a change via Hot Module Replacement (HMR) in
a hybrid app would result in an error. This was caused by not having the
AngularJS app destroyed and thus trying to bootstrap an AngularJS app on
the same element twice.

This commit fixes HMR for hybrid apps by ensuring the AngularJS app is
destroyed when the Angular `PlatformRef` is [destroyed][1] in the
[`module.hot.dispose()` callback][2].

NOTE:
For "ngUpgradeLite" apps (i.e. those using `downgradeModule()`), HMR
will only work if the downgraded module has been bootstrapped and there
is at least one Angular component present on the page. The is due to a
combination of two facts:
- The logic for setting up the listener that destroys the AngularJS app
  depends on the downgraded module's `NgModuleRef`, which is only
  available after the module has been bootstrapped.
- The [HMR dispose logic][3] depends on having an Angular element
  (identified by the auto-geenrated `ng-version` attribute) present in
  the DOM in order to retrieve the Angular `PlatformRef`.

[1]:
https://github.com/angular/angular-cli/blob/205ea2b638f154291993bfd9e065cd66ff20503/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L75
[2]:
https://github.com/angular/angular-cli/blob/205ea2b638f154291993bfd9e065cd66ff205033/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L31
[3]:
https://github.com/angular/angular-cli/blob/205ea2b638f154291993bfd9e065cd66ff205033/packages/angular_devkit/build_angular/src/webpack/plugins/hmr/hmr-accept.ts#L116

Fixes #39935

gkalpak committed:
fixup! fix(upgrade): fix HMR for hybrid applications

gkalpak commented on packages/upgrade/src/dynamic/src/upgrade_adapter.ts:
```diff
@@ -620,6 +619,12 @@ export class UpgradeAdapter {
+                    // Destroy the AngularJS app once the Angular `PlatformRef` is destroyed.
+                    // This does not happen in a typical SPA scenario, but it might be useful for
+                    // other usecases where desposing of an Angular/AngularJS app is necessary (such
+                    // as Hot Module Replacement (HMR)).
```
Done.

gkalpak commented on packages/upgrade/src/common/src/util.ts:
```diff
@@ -25,10 +25,36 @@ export function onError(e: any) {
 
+export function cleanData(node: Node): void {
+  // Clean the jqLite/jQuery data on the element and all its descendants.
+  // Equivalent to how jqLite/jQuery invoke `cleanData()` on an Element when removed:
```
Done.

gkalpak commented:
>> For "ngUpgradeLite" apps (i.e. those using downgradeModule()), HMR will only work if there is at least one Angular component present on the page.
>
>Can you explain why this is the case?

Expanded on this in the commit message. (Also updated the PR description.)

##
Thx for the reviews. I have addressed the comments. Marking for merging...

mary-poppins commented:
You can preview 885710a at https://pr40045-885710a.ngbuilds.io/.
